File: parse-bedmesh.py
#!/usr/bin/env python3

import glob
import re
from collections import defaultdict

pattern = re.compile(r"Bed X:\s*([-\d.]+)\s*Y:\s*([-\d.]+)\s*Z:\s*([-\d.]+)")

points = []

# Read all matching files
for filename in glob.glob("gcode_G29_P10_V4_*"):
    with open(filename, "r") as f:
        for line in f:
            m = pattern.search(line)
            if m:
                x = float(m.group(1))
                y = float(m.group(2))
                z = float(m.group(3))
                points.append((x, y, z))

# Collect sorted unique axes
xs = sorted(set(p[0] for p in points))
ys = reversed(sorted(set(p[1] for p in points)))

# Build lookup
grid = {(x, y): z for (x, y, z) in points}

# Header
header = "".join(f"x={int(x):<6}" for x in xs)

print("MESH_Z = [")
print(f"  # {header}(for each y row)")

# Rows
for y in ys:
    row_vals = []
    for x in xs:
        z = grid.get((x, y), 0.0)
        row_vals.append(f"{z: .3f}")
    row_str = ",  ".join(row_vals)
    print(f"  {row_str},  # y={int(y)}")

print("].freeze")

File: parse-bedmesh.rb
#!/usr/bin/env ruby

files = Dir.glob("gcode_G29_P10_V4_*")

points = []

# Extract data
files.each do |file|
  File.foreach(file) do |line|
    if line =~ /Bed X:\s*([-\d.]+)\s*Y:\s*([-\d.]+)\s*Z:\s*([-\d.]+)/
      x = $1.to_f
      y = $2.to_f
      z = $3.to_f
      points << [x, y, z]
    end
  end
end

# Get unique sorted axes
xs = points.map { |p| p[0] }.uniq.sort
ys = points.map { |p| p[1] }.uniq.sort.reverse

# Build lookup hash
grid = {}
points.each do |x, y, z|
  grid[[x, y]] = z
end

# Header
header = xs.map { |x| "x=#{x.to_i}".ljust(8) }.join
puts "MESH_Z = ["
puts "  # #{header}(for each y row)"

# Rows
ys.each do |y|
  row = xs.map do |x|
    val = grid[[x, y]] || 0.0
    "% .3f" % val
  end.join(",  ")

  puts "  #{row},  # y=#{y.to_i}"
end

puts "].freeze"
